pyro/distributions/hmm.py

```diff
@@ -0,0 +1,95 @@
+from __future__ import absolute_import, division, print_function
+
+import torch
+from torch.distributions import constraints
+
+from pyro.distributions.torch_distribution import TorchDistribution
+from pyro.distributions.util import broadcast_shape
+
+
+def _logmatmulexp(x, y):
+    """
+    Numerically stable version of ``(x.log() @ y.log()).exp()``.
+    """
+    x_shift = x.max(-1, keepdim=True)[0]
+    y_shift = y.max(-2, keepdim=True)[0]
+    xy = torch.matmul((x - x_shift).exp(), (y - y_shift).exp()).log()
+    return xy + x_shift + y_shift
+
+
+def _sequential_logmatmulexp(logits):
+    """
+    For a tensor ``x`` whose time dimension is -3, computes::
+
+        x[..., 0, :, :] @ x[..., 1, :, :] @ ... @ x[..., T-1, :, :]
+
+    but does so numerically stably in log space.
+    """
+    batch_shape = logits.shape[:-3]
+    state_dim = logits.size(-1)
+    while logits.size(-3) > 1:
+        time = logits.size(-3)
+        even_time = time // 2 * 2
+        even_part = logits[..., :even_time, :, :]
+        x_y = even_part.reshape(batch_shape + (even_time // 2, 2, state_dim, state_dim))
+        x, y = x_y.unbind(-3)
+        contracted = _logmatmulexp(x, y)
```

Comment:
Note if `x`, `y` were `funsor.Gaussian`s, we would be able to reuse this code for Gaussian HMMs simply by changing the definition of this operation

Reply:
Yeah I'd like to gradually make more abstract versions of this code, including multiple `opt_einsum` backends, more complex model structure (e.g. passing in an einsum eqution), and eventually Funsor support. In this first PR I've tried to customize so we can measure speedup with minimal abstraction overhead.